core/src/main/scala/kafka/tools/MirrorMaker.scala

```diff
@@ -78,120 +82,202 @@ object MirrorMaker extends Logging with KafkaMetricsGroup {
-    producer = new MirrorMakerProducer(producerProps)
 
+      CommandLineUtils.checkRequiredArgs(parser, options, consumerConfigOpt, producerConfigOpt)
+      if (List(whitelistOpt, blacklistOpt).count(options.has) != 1) {
+        println("Exactly one of whitelist or blacklist is required.")
+        System.exit(1)
+      }
+
+      abortOnSendFailure = options.valueOf(abortOnSendFailureOpt).toBoolean
+      offsetCommitIntervalMs = options.valueOf(offsetCommitIntervalMsOpt).intValue()
+      val numStreams = options.valueOf(numStreamsOpt).intValue()
+
+      Runtime.getRuntime.addShutdownHook(new Thread("MirrorMakerShutdownHook") {
+        override def run() {
+          cleanShutdown()
+        }
+      })
+
+      // create producer
+      val producerProps = Utils.loadProps(options.valueOf(producerConfigOpt))
+      // Defaults to no data loss settings.
+      maybeSetDefaultProperty(producerProps, ProducerConfig.RETRIES_CONFIG, Int.MaxValue.toString)
+      maybeSetDefaultProperty(producerProps, ProducerConfig.BLOCK_ON_BUFFER_FULL_CONFIG, "true")
+      maybeSetDefaultProperty(producerProps, ProducerConfig.ACKS_CONFIG, "all")
+      maybeSetDefaultProperty(producerProps, ProducerConfig.MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION, "1")
+      // Always set producer key and value serializer to ByteArraySerializer.
+      producerProps.setProperty(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArraySerializer")
+      producerProps.setProperty(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArraySerializer")
+      producer = new MirrorMakerProducer(producerProps)
+
+      val useNewConsumer = options.has(useNewConsumerOpt)
+
+      // Set consumer rebalance listener.
+      // Custom rebalance listener will be invoked after internal listener finishes its work.
+      val customRebalanceListener = {
+        val customRebalanceListenerClass = options.valueOf(consumerRebalanceListenerOpt)
+        if (customRebalanceListenerClass != null) {
+          val rebalanceListenerArgs = options.valueOf(rebalanceListenerArgsOpt)
+          if (rebalanceListenerArgs != null) {
+            if (useNewConsumer)
+              Some(CoreUtils.createObject[org.apache.kafka.clients.consumer.ConsumerRebalanceListener](customRebalanceListenerClass, rebalanceListenerArgs))
+            else
+              Some(CoreUtils.createObject[ConsumerRebalanceListener](customRebalanceListenerClass, rebalanceListenerArgs))
+          } else {
+            if (useNewConsumer)
+              Some(CoreUtils.createObject[org.apache.kafka.clients.consumer.ConsumerRebalanceListener](customRebalanceListenerClass))
+            else
+              Some(CoreUtils.createObject[ConsumerRebalanceListener](customRebalanceListenerClass))
+          }
+        } else {
+          None
+        }
+      }
+
+      // Create consumers
+      val mirrorMakerConsumers = if (!useNewConsumer) {
```

Comment:
We can collapse the creation of the consumer and the listener together here with

```
if (useNewConsumer)
    if (class != null)
        if (args != null)
        else
    else
else
    if (class != null)
        if (args != null)
        else
    else
```

And relaying on cast exceptions for incorrect matching of consumer types.